lib/value/src/kind/comparison.rs

```diff
@@ -13,6 +13,7 @@ impl Kind {
             && self.contains_timestamp()
             && self.contains_regex()
             && self.contains_null()
+            && self.contains_undefined()
```

Comment:
Are we sure we want `any` to be defined as "any type, including not being defined at all"?

It almost feels like modelling this properly would require two levels, first you'd have an `undefined` vs `defined` state for a `Kind`, and then for defined ones, you'd have all the different type states that kind can have.

That is, there are two separate questions you'd want to be able to ask about the state of a value:

1. Do we know if the value can exist at runtime?
2. If it can exist, what do we know about the potential types of the value?

In this implementation, we're merging those two questions, which (it seems to me, but I might be off here) muddies the water when it comes to methods like `any`, which no longer implies "any type", but "any type _and_ the potential for the value to not exist at all".

I haven't thought this through thoroughly yet, but this _feels_ like it would complicate the reasoning we need to do on type definitions.

I'm curious about your thoughts here, and also happy to chat offline.

Reply:
> Are we sure we want any to be defined as "any type, including not being defined at all"?

Yes. `any` means we have no type information (it can be anything). If you know it can be anything except undefined, you can set it to `Kind::any().without_undefined()`.
I think you are assuming `undefined` is more special than it is. It's just another type a value can be, and happens to be the default type for values not defined.